File: proto2ros_tests/proto2ros_tests/manual_conversions.py
import math

import bosdyn.api.geometry_pb2
import geometry_msgs.msg

from proto2ros.conversions import convert


@convert.register(geometry_msgs.msg.Vector3, bosdyn.api.geometry_pb2.Vec3)
def convert_geometry_msgs_vector3_message_to_bosdyn_api_vec3_proto(
    ros_msg: geometry_msgs.msg.Vector3, proto_msg: bosdyn.api.geometry_pb2.Vec3
) -> None:
    proto_msg.x = ros_msg.x
    proto_msg.y = ros_msg.y
    proto_msg.z = ros_msg.z


@convert.register(bosdyn.api.geometry_pb2.Vec3, geometry_msgs.msg.Vector3)
def convert_bosdyn_api_vec3_proto_to_geometry_msgs_vector3_message(
    proto_msg: bosdyn.api.geometry_pb2.Vec3, ros_msg: geometry_msgs.msg.Vector3
) -> None:
    ros_msg.x = proto_msg.x
    ros_msg.y = proto_msg.y
    ros_msg.z = proto_msg.z


@convert.register(geometry_msgs.msg.Quaternion, bosdyn.api.geometry_pb2.Quaternion)
def convert_geometry_msgs_quaternion_message_to_bosdyn_api_quaternion_proto(
    ros_msg: geometry_msgs.msg.Quaternion, proto_msg: bosdyn.api.geometry_pb2.Quaternion
) -> None:
    proto_msg.w = ros_msg.w
    proto_msg.x = ros_msg.x
    proto_msg.y = ros_msg.y
    proto_msg.z = ros_msg.z


@convert.register(bosdyn.api.geometry_pb2.Quaternion, geometry_msgs.msg.Quaternion)
def convert_bosdyn_api_quaternion_proto_to_geometry_msgs_quaternion_message(
    proto_msg: bosdyn.api.geometry_pb2.Quaternion, ros_msg: geometry_msgs.msg.Quaternion
) -> None:
    ros_msg.w = proto_msg.w
    ros_msg.x = proto_msg.x
    ros_msg.y = proto_msg.y
    ros_msg.z = proto_msg.z


@convert.register(geometry_msgs.msg.Pose, bosdyn.api.geometry_pb2.SE3Pose)
def convert_geometry_msgs_pose_message_to_bosdyn_api_se3_pose_proto(
    ros_msg: geometry_msgs.msg.Pose, proto_msg: bosdyn.api.geometry_pb2.SE3Pose
) -> None:
    convert_geometry_msgs_vector3_message_to_bosdyn_api_vec3_proto(ros_msg.position, proto_msg.position)
    convert_geometry_msgs_quaternion_message_to_bosdyn_api_quaternion_proto(ros_msg.orientation, proto_msg.rotation)


@convert.register(bosdyn.api.geometry_pb2.SE3Pose, geometry_msgs.msg.Pose)
def convert_bosdyn_api_se3_pose_proto_to_geometry_msgs_pose_message(
    proto_msg: bosdyn.api.geometry_pb2.SE3Pose, ros_msg: geometry_msgs.msg.Pose
) -> None:
    convert_bosdyn_api_vec3_proto_to_geometry_msgs_vector3_message(proto_msg.position, ros_msg.position)
    convert_bosdyn_api_quaternion_proto_to_geometry_msgs_quaternion_message(proto_msg.rotation, ros_msg.orientation)


@convert.register(geometry_msgs.msg.Pose, bosdyn.api.geometry_pb2.SE2Pose)
def convert_geometry_msgs_pose_message_to_bosdyn_api_se2_pose_proto(
    ros_msg: geometry_msgs.msg.Pose, proto_msg: bosdyn.api.geometry_pb2.SE2Pose
) -> None:
    proto_msg.position.x = ros_msg.position.x
    proto_msg.position.y = ros_msg.position.y
    proto_msg.angle = 2.0 * math.acos(ros_msg.orientation.w)


@convert.register(bosdyn.api.geometry_pb2.SE2Pose, geometry_msgs.msg.Pose)
def convert_bosdyn_api_se2_pose_proto_to_geometry_msgs_pose_message(
    proto_msg: bosdyn.api.geometry_pb2.SE2Pose, ros_msg: geometry_msgs.msg.Pose
) -> None:
    ros_msg.position.x = proto_msg.position.x
    ros_msg.position.y = proto_msg.position.y
    ros_msg.orientation.w = math.cos(proto_msg.angle / 2.0)
    ros_msg.orientation.z = math.sin(proto_msg.angle / 2.0)


@convert.register(geometry_msgs.msg.Polygon, bosdyn.api.geometry_pb2.Polygon)
def convert_geometry_msgs_polygon_message_to_bosdyn_api_polygon_proto(
    ros_msg: geometry_msgs.msg.Polygon, proto_msg: bosdyn.api.geometry_pb2.Polygon
) -> None:
    proto_msg.Clear()
    for point in ros_msg.points:
        vertex = proto_msg.vertexes.add()
        vertex.x = point.x
        vertex.y = point.y


@convert.register(bosdyn.api.geometry_pb2.Polygon, geometry_msgs.msg.Polygon)
def convert_bosdyn_api_polygon_proto_to_geometry_msgs_polygon_message(
    proto_msg: bosdyn.api.geometry_pb2.Polygon, ros_msg: geometry_msgs.msg.Polygon
) -> None:
    ros_msg.points = [geometry_msgs.msg.Point32(x=vertex.x, y=vertex.y, z=0.0) for vertex in proto_msg.vertexes]


@convert.register(geometry_msgs.msg.Vector3, bosdyn.api.geometry_pb2.Circle)
def convert_geometry_msgs_vector3_message_to_bosdyn_api_circle_proto(
    ros_msg: geometry_msgs.msg.Vector3, proto_msg: bosdyn.api.geometry_pb2.Circle
) -> None:
    proto_msg.center_pt.x = ros_msg.x
    proto_msg.center_pt.y = ros_msg.y
    proto_msg.radius = ros_msg.z


@convert.register(bosdyn.api.geometry_pb2.Circle, geometry_msgs.msg.Vector3)
def convert_bosdyn_api_circle_proto_to_geometry_msgs_vector3_message(
    proto_msg: bosdyn.api.geometry_pb2.Circle, ros_msg: geometry_msgs.msg.Vector3
) -> None:
    ros_msg.x = proto_msg.center_pt.x
    ros_msg.y = proto_msg.center_pt.y
    ros_msg.z = proto_msg.radius


@convert.register(geometry_msgs.msg.Twist, bosdyn.api.geometry_pb2.SE3Velocity)
def convert_geometry_msgs_twist_message_to_bosdyn_api_se3_velocity_proto(
    ros_msg: geometry_msgs.msg.Twist, proto_msg: bosdyn.api.geometry_pb2.SE3Velocity
) -> None:
    convert_geometry_msgs_vector3_message_to_bosdyn_api_vec3_proto(ros_msg.linear, proto_msg.linear)
    convert_geometry_msgs_vector3_message_to_bosdyn_api_vec3_proto(ros_msg.angular, proto_msg.angular)


@convert.register(bosdyn.api.geometry_pb2.SE3Velocity, geometry_msgs.msg.Twist)
def convert_bosdyn_api_se3_velocity_proto_to_geometry_msgs_twist_message(
    proto_msg: bosdyn.api.geometry_pb2.SE3Velocity, ros_msg: geometry_msgs.msg.Twist
) -> None:
    convert_bosdyn_api_vec3_proto_to_geometry_msgs_vector3_message(proto_msg.linear, ros_msg.linear)
    convert_bosdyn_api_vec3_proto_to_geometry_msgs_vector3_message(proto_msg.angular, ros_msg.angular)


@convert.register(geometry_msgs.msg.Wrench, bosdyn.api.geometry_pb2.Wrench)
def convert_geometry_msgs_wrench_message_to_bosdyn_api_wrench_proto(
    ros_msg: geometry_msgs.msg.Wrench, proto_msg: bosdyn.api.geometry_pb2.Wrench
) -> None:
    convert_geometry_msgs_vector3_message_to_bosdyn_api_vec3_proto(ros_msg.force, proto_msg.force)
    convert_geometry_msgs_vector3_message_to_bosdyn_api_vec3_proto(ros_msg.torque, proto_msg.torque)


@convert.register(bosdyn.api.geometry_pb2.Wrench, geometry_msgs.msg.Wrench)
def convert_bosdyn_api_wrench_proto_to_geometry_msgs_wrench_messagey(
    proto_msg: bosdyn.api.geometry_pb2.Wrench, ros_msg: geometry_msgs.msg.Wrench
) -> None:
    convert_bosdyn_api_vec3_proto_to_geometry_msgs_vector3_message(proto_msg.force, ros_msg.force)
    convert_bosdyn_api_vec3_proto_to_geometry_msgs_vector3_message(proto_msg.torque, ros_msg.torque)

File: proto2ros_tests/src/manual_conversions.cpp
// Copyright (c) 2024 Boston Dynamics AI Institute LLC. All rights reserved.

#include "proto2ros_tests/manual_conversions.hpp"

#include <cmath>

namespace proto2ros_tests::conversions {

void Convert(const geometry_msgs::msg::Vector3& ros_msg, bosdyn::api::Vec3* proto_msg) {
  proto_msg->set_x(ros_msg.x);
  proto_msg->set_y(ros_msg.y);
  proto_msg->set_z(ros_msg.z);
}

void Convert(const bosdyn::api::Vec3& proto_msg, geometry_msgs::msg::Vector3* ros_msg) {
  ros_msg->x = proto_msg.x();
  ros_msg->y = proto_msg.y();
  ros_msg->z = proto_msg.z();
}

void Convert(const geometry_msgs::msg::Point& ros_msg, bosdyn::api::Vec3* proto_msg) {
  proto_msg->set_x(ros_msg.x);
  proto_msg->set_y(ros_msg.y);
  proto_msg->set_z(ros_msg.z);
}

void Convert(const bosdyn::api::Vec3& proto_msg, geometry_msgs::msg::Point* ros_msg) {
  ros_msg->x = proto_msg.x();
  ros_msg->y = proto_msg.y();
  ros_msg->z = proto_msg.z();
}

void Convert(const geometry_msgs::msg::Quaternion& ros_msg, bosdyn::api::Quaternion* proto_msg) {
  proto_msg->set_x(ros_msg.x);
  proto_msg->set_y(ros_msg.y);
  proto_msg->set_z(ros_msg.z);
  proto_msg->set_w(ros_msg.w);
}

void Convert(const bosdyn::api::Quaternion& proto_msg, geometry_msgs::msg::Quaternion* ros_msg) {
  ros_msg->x = proto_msg.x();
  ros_msg->y = proto_msg.y();
  ros_msg->z = proto_msg.z();
  ros_msg->w = proto_msg.w();
}

void Convert(const geometry_msgs::msg::Pose& ros_msg, bosdyn::api::SE3Pose* proto_msg) {
  Convert(ros_msg.position, proto_msg->mutable_position());
  Convert(ros_msg.orientation, proto_msg->mutable_rotation());
}

void Convert(const bosdyn::api::SE3Pose& proto_msg, geometry_msgs::msg::Pose* ros_msg) {
  Convert(proto_msg.position(), &ros_msg->position);
  Convert(proto_msg.rotation(), &ros_msg->orientation);
}

void Convert(const geometry_msgs::msg::Pose& ros_msg, bosdyn::api::SE2Pose* proto_msg) {
  // NOLINTBEGIN(readability-magic-numbers)
  proto_msg->Clear();
  proto_msg->mutable_position()->set_x(ros_msg.position.x);
  proto_msg->mutable_position()->set_y(ros_msg.position.y);
  proto_msg->set_angle(2.0 * std::acos(ros_msg.orientation.w));
  const double angle = 2.0 * std::asin(ros_msg.orientation.z);
  if (std::abs(proto_msg->angle() - angle) > std::numeric_limits<double>::epsilon()) {
    throw std::domain_error("geometry_msgs/Pose does not represent a bosdyn.api.SE2Pose");
  }
  // NOLINTEND(readability-magic-numbers)
}

void Convert(const bosdyn::api::SE2Pose& proto_msg, geometry_msgs::msg::Pose* ros_msg) {
  // NOLINTBEGIN(readability-magic-numbers)
  ros_msg->position.x = proto_msg.position().x();
  ros_msg->position.y = proto_msg.position().y();
  ros_msg->position.z = 0.0;
  ros_msg->orientation.w = std::cos(proto_msg.angle() / 2.0);
  ros_msg->orientation.z = std::sin(proto_msg.angle() / 2.0);
  ros_msg->orientation.y = 0.0;
  ros_msg->orientation.x = 0.0;
  // NOLINTEND(readability-magic-numbers)
}

void Convert(const geometry_msgs::msg::Polygon& ros_msg, bosdyn::api::Polygon* proto_msg) {
  proto_msg->Clear();
  auto* vertices = proto_msg->mutable_vertexes();
  vertices->Reserve(static_cast<int>(ros_msg.points.size()));
  for (const auto& point : ros_msg.points) {
    auto* vertex = vertices->Add();
    vertex->set_x(point.x);
    vertex->set_y(point.y);
  }
}

void Convert(const bosdyn::api::Polygon& proto_msg, geometry_msgs::msg::Polygon* ros_msg) {
  ros_msg->points.clear();
  ros_msg->points.reserve(proto_msg.vertexes_size());
  for (const auto& vertex : proto_msg.vertexes()) {
    auto& point = ros_msg->points.emplace_back();
    point.x = static_cast<float>(vertex.x());
    point.y = static_cast<float>(vertex.y());
  }
}

void Convert(const geometry_msgs::msg::Vector3& ros_msg, bosdyn::api::Circle* proto_msg) {
  auto* center = proto_msg->mutable_center_pt();
  center->set_x(ros_msg.x);
  center->set_y(ros_msg.y);
  proto_msg->set_radius(ros_msg.z);
}

void Convert(const bosdyn::api::Circle& proto_msg, geometry_msgs::msg::Vector3* ros_msg) {
  const auto& center = proto_msg.center_pt();
  ros_msg->x = center.x();
  ros_msg->y = center.y();
  ros_msg->z = proto_msg.radius();
}

void Convert(const geometry_msgs::msg::Twist& ros_msg, bosdyn::api::SE3Velocity* proto_msg) {
  Convert(ros_msg.linear, proto_msg->mutable_linear());
  Convert(ros_msg.angular, proto_msg->mutable_angular());
}

void Convert(const bosdyn::api::SE3Velocity& proto_msg, geometry_msgs::msg::Twist* ros_msg) {
  Convert(proto_msg.linear(), &ros_msg->linear);
  Convert(proto_msg.angular(), &ros_msg->angular);
}

void Convert(const geometry_msgs::msg::Wrench& ros_msg, bosdyn::api::Wrench* proto_msg) {
  Convert(ros_msg.force, proto_msg->mutable_force());
  Convert(ros_msg.torque, proto_msg->mutable_torque());
}

void Convert(const bosdyn::api::Wrench& proto_msg, geometry_msgs::msg::Wrench* ros_msg) {
  Convert(proto_msg.force(), &ros_msg->force);
  Convert(proto_msg.torque(), &ros_msg->torque);
}

void Convert(const sensor_msgs::msg::Temperature& ros_msg, Temperature* proto_msg) {
  proto_msg->set_scale(Temperature::CELSIUS);
  proto_msg->set_value(static_cast<float>(ros_msg.temperature));
}

void Convert(const Temperature& proto_msg, sensor_msgs::msg::Temperature* ros_msg) {
  // NOLINTBEGIN(readability-magic-numbers)
  switch (proto_msg.scale()) {
    case Temperature::KELVIN:
      ros_msg->temperature = proto_msg.value() + 273.0;
      break;
    case Temperature::FAHRENHEIT:
      ros_msg->temperature = (proto_msg.value() - 32.0) * 5.0 / 9.0;
      break;
    case Temperature::CELSIUS:
    default:
      ros_msg->temperature = proto_msg.value();
      break;
  }
  // NOLINTEND(readability-magic-numbers)
}

}  // namespace proto2ros_tests::conversions
